### PR TITLE
Pull request for the authentication project.

### PR DESCRIPTION
**Pull request by RoodzFernando**
We have implemented the things below.
- A user model and the controllers
- Add the different views and forms
- Add a signup and login forms. 
- A post model
- Add the authorization required for a logged user.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  validates :password, presence: true, length: { minimum: 6 }
+
+  # Returns the hash digest of the given string.
+  def self.digest(string)
+    cost = ActiveModel::SecurePassword.min_cost unless BCrypt::Engine::MIN_COST
+
+    BCrypt::Engine.cost
+
+    BCrypt::Password.create(string, cost: cost)
+  end
```
Comment (maelfosso) [inline app/models/user.rb]: Not really needed. Use `Digest::SHA1.hexdigest` which is required in project specification

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
+module SessionsHelper
+  # logs in the given user.
+  def log_in(user)
+    session[:user_id] = user.id
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: Forget about using `session`. Only `cookies` is needed.

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  def current_user
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user&.authenticated?(cookies[:remember_token])
+        log_in user
+        @current_user = user
+      end
+    end
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: It's too long.
Kindly, consider only `cookies` and check whether or not an user with the `cookies.permanent[:remember_token]` exists

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  # Returns true if the user is logged in, false otherwise.
+  def logged_in?
+    !current_user.nil?
+  end
+
+  def forget(_user)
+    # user.forget
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: Remove comment

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello @RoodzFernando @firmoholanda 

Well done job! IT's correctly working but there are some requirements that need to be met.

1. As indicated [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_create_requirement.png), It has a `#before_create` callback that generates a :remember_token 

2. Don't think about using `session`. The `cookies` is sufficient.

3. Please, fix all the stickler's errors.

4. Read the comment below.

Kindly make the suggested changes and submit a new code review request.
Thanks

Mael FOSSO

**Review by mcrd25 — APPROVED** [review]
# Changes Approved! 💪 🤓 🎊 

Hi [Holanda](@firmoholanda )! Hi [Fernando](@RoodzFernando )! 👋 

Good job with the changes! I see no more for further review! 😄 

I do however have some optional suggestions.

#### Additional Suggestions
- Currently, a user can edit another user's info. Maybe this should not be the case [OPTIONAL]
- The create post button should be seen right away, I almost missed it [OPTIONAL]
- Sign up button should not be shown when a user is logged in [OPTIONAL]

***If you have any questions, now or in the future, do not hesitate to contact me on social media, email or on slack and I will be glad to chat with you.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

**Pull request merged**